ClientScheduledExecutorServiceBasicTest.cancelledAndDone_durable

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.scheduledexecutor.ScheduledExecutorServiceBasicTest.cancelledAndDone_durable(ScheduledExecutorServiceBasicTest.java:529)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:89)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:81)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.lang.Thread.run(Thread.java:662)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance-OracleJDK1.6-freestyle/10/testReport/junit/com.hazelcast.client.scheduledexecutor/ClientScheduledExecutorServiceBasicTest/cancelledAndDone_durable/

--- a/hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorContainer.java
+++ b/hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorContainer.java
@@ -201,7 +201,17 @@ public class ScheduledExecutorContainer {
 
         descriptor.setState(newState);
         descriptor.setStats(stats);
-        descriptor.setTaskResult(resolution);
+
+        if (descriptor.getTaskResult() != null) {
+            // Task result previously populated - ie. through cancel
+            // Ignore all subsequent results
+            if (logger.isFineEnabled()) {
+                logger.fine(String.format("[Scheduler: " + name + "][Partition: " + partitionId + "] syncState result skipped. "
+                                        + "Current: %s New: %s", descriptor.getTaskResult(), resolution));
+            }
+        } else {
+            descriptor.setTaskResult(resolution);
+        }
     }
 
     public boolean shouldParkGetResult(String taskName)

--- a/hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/operations/SyncStateOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/operations/SyncStateOperation.java
@@ -41,7 +41,7 @@ public class SyncStateOperation
 
     protected ScheduledTaskResult result;
 
-    protected boolean shouldRun;
+    private boolean shouldRun;
 
     public SyncStateOperation() {
     }

--- a/hazelcast/src/test/java/com/hazelcast/scheduledexecutor/ScheduledExecutorServiceSlowTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/scheduledexecutor/ScheduledExecutorServiceSlowTest.java
@@ -31,6 +31,8 @@ import java.util.concurrent.ExecutionException;
 
 import static java.util.concurrent.TimeUnit.SECONDS;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({SlowTest.class, ParallelTest.class})
@@ -50,7 +52,7 @@ public class ScheduledExecutorServiceSlowTest
 
         IScheduledExecutorService executorService = getScheduledExecutor(instances, "s");
         IScheduledFuture<Double> future = executorService.schedule(
-                new ScheduledExecutorServiceBasicTest.ICountdownLatchCallableTask("runsCountLatchName", 15000), delay, SECONDS);
+                new ICountdownLatchCallableTask("runsCountLatchName", 15000), delay, SECONDS);
 
         double result = future.get();
 
@@ -75,7 +77,7 @@ public class ScheduledExecutorServiceSlowTest
         latch.trySetCount(1);
 
         IScheduledFuture future = executorService.scheduleOnKeyOwnerAtFixedRate(
-                new ScheduledExecutorServiceBasicTest.StatefulRunnableTask("latch", "runC", "loadC"),
+                new StatefulRunnableTask("latch", "runC", "loadC"),
                 key, 10, 10, SECONDS);
 
         // Wait for task to get scheduled and start
@@ -107,7 +109,7 @@ public class ScheduledExecutorServiceSlowTest
 
         IScheduledExecutorService executorService = getScheduledExecutor(instances, "s");
         IScheduledFuture future = executorService.scheduleOnKeyOwnerAtFixedRate(
-                new ScheduledExecutorServiceBasicTest.ICountdownLatchRunnableTask("latch"), key, 0, 10, SECONDS);
+                new ICountdownLatchRunnableTask("latch"), key, 0, 10, SECONDS);
 
         Thread.sleep(12000);
 
@@ -131,7 +133,7 @@ public class ScheduledExecutorServiceSlowTest
 
         IScheduledExecutorService executorService = getScheduledExecutor(instances, "s");
         IScheduledFuture future = executorService.scheduleAtFixedRate(
-                new ScheduledExecutorServiceBasicTest.ICountdownLatchRunnableTask("latch"), 0, 10, SECONDS);
+                new ICountdownLatchRunnableTask("latch"), 0, 10, SECONDS);
 
 
         latch.await(120, SECONDS);
@@ -293,4 +295,77 @@ public class ScheduledExecutorServiceSlowTest
 
         assertEquals(expectedTotal, actualTotal, 0);
     }
+
+    @Test
+    public void cancelUninterruptedTask_waitUntilRunCompleted_checkStatusIsCancelled()
+            throws ExecutionException, InterruptedException {
+
+        HazelcastInstance[] instances = createClusterWithCount(1);
+
+        String runFinishedLatchName = "runFinishedLatch";
+        ICountDownLatch latch = instances[0].getCountDownLatch(runFinishedLatchName);
+        latch.trySetCount(1);
+
+        IScheduledExecutorService executorService = getScheduledExecutor(instances, "s");
+        IScheduledFuture future = executorService.scheduleAtFixedRate(
+                new HotLoopBusyTask(runFinishedLatchName), 0, 1, SECONDS);
+
+        assertFalse(future.isCancelled());
+        assertFalse(future.isDone());
+
+        future.cancel(false);
+
+        assertTrue(future.isCancelled());
+        assertTrue(future.isDone());
+
+        // Even though we cancelled the task is current task is still running.
+        // Wait till the task is actually done
+        latch.await(60, SECONDS);
+
+        // Make sure SyncState goes through
+        sleepSeconds(10);
+
+        // Check once more that the task status is consistent
+        assertTrue(future.isCancelled());
+        assertTrue(future.isDone());
+    }
+
+    @Test
+    public void cancelUninterruptedTask_waitUntilRunCompleted_killMember_checkStatusIsCancelled()
+            throws ExecutionException, InterruptedException {
+
+        HazelcastInstance[] instances = createClusterWithCount(2);
+
+        String key = generateKeyOwnedBy(instances[1]);
+
+        String runFinishedLatchName = "runFinishedLatch";
+        ICountDownLatch latch = instances[0].getCountDownLatch(runFinishedLatchName);
+        latch.trySetCount(1);
+
+        IScheduledExecutorService executorService = getScheduledExecutor(instances, "s");
+        IScheduledFuture future = executorService.scheduleOnKeyOwnerAtFixedRate(
+                new HotLoopBusyTask(runFinishedLatchName), key, 0, 1, SECONDS);
+
+        assertFalse(future.isCancelled());
+        assertFalse(future.isDone());
+
+        future.cancel(false);
+
+        assertTrue(future.isCancelled());
+        assertTrue(future.isDone());
+
+        // Even though we cancelled the task is current task is still running.
+        // Wait till the task is actually done
+        latch.await(60, SECONDS);
+
+        // Make sure SyncState goes through
+        sleepSeconds(10);
+
+        instances[1].getLifecycleService().terminate();
+
+        // Check once more that the task status is consistent
+        assertTrue(future.isCancelled());
+        assertTrue(future.isDone());
+    }
+
 }

--- a/hazelcast/src/test/java/com/hazelcast/scheduledexecutor/ScheduledExecutorServiceTestSupport.java
+++ b/hazelcast/src/test/java/com/hazelcast/scheduledexecutor/ScheduledExecutorServiceTestSupport.java
@@ -31,6 +31,9 @@ import java.util.List;
 import java.util.Map;
 import java.util.concurrent.Callable;
 
+import static java.lang.System.currentTimeMillis;
+import static java.lang.Thread.sleep;
+
 /**
  * Common methods used in ScheduledExecutorService tests
  */
@@ -127,7 +130,7 @@ public class ScheduledExecutorServiceTestSupport extends HazelcastTestSupport {
         @Override
         public Double call() {
             try {
-                Thread.sleep(sleepPeriod);
+                sleep(sleepPeriod);
             } catch (InterruptedException e) {
                 Thread.interrupted();
             }
@@ -180,7 +183,8 @@ public class ScheduledExecutorServiceTestSupport extends HazelcastTestSupport {
         }
     }
 
-    static class ICountdownLatchRunnableTask implements Runnable, Serializable, HazelcastInstanceAware {
+    static class ICountdownLatchRunnableTask
+            implements Runnable, Serializable, HazelcastInstanceAware {
 
         final String runsCountlatchName;
 
@@ -201,7 +205,40 @@ public class ScheduledExecutorServiceTestSupport extends HazelcastTestSupport {
         }
     }
 
-    static class PlainCallableTask implements Callable<Double>, Serializable {
+    static class HotLoopBusyTask
+            implements Runnable, HazelcastInstanceAware, Serializable {
+
+        private final String runFinishedLatchName;
+        private transient HazelcastInstance instance;
+
+        HotLoopBusyTask(String runFinishedLatchName) {
+            this.runFinishedLatchName = runFinishedLatchName;
+        }
+
+        @Override
+        public void run() {
+            long start = currentTimeMillis();
+            while (true) {
+                try {
+                    sleep(5000);
+                    if (currentTimeMillis() - start >= 30000) {
+                        instance.getCountDownLatch(runFinishedLatchName).countDown();
+                        break;
+                    }
+                } catch (InterruptedException e) {
+                    // ignore
+                }
+            }
+        }
+
+        @Override
+        public void setHazelcastInstance(HazelcastInstance hazelcastInstance) {
+            this.instance = hazelcastInstance;
+        }
+    }
+
+    static class PlainCallableTask
+            implements Callable<Double>, Serializable {
 
         private int delta = 0;
 
@@ -220,7 +257,8 @@ public class ScheduledExecutorServiceTestSupport extends HazelcastTestSupport {
 
     }
 
-    static class ErroneousCallableTask implements Callable<Double>, Serializable, HazelcastInstanceAware {
+    static class ErroneousCallableTask
+            implements Callable<Double>, Serializable, HazelcastInstanceAware {
 
         private String completionLatchName;
 
@@ -250,7 +288,8 @@ public class ScheduledExecutorServiceTestSupport extends HazelcastTestSupport {
         }
     }
 
-    static class PlainPartitionAwareCallableTask implements Callable<Double>, Serializable, PartitionAware<String> {
+    static class PlainPartitionAwareCallableTask
+            implements Callable<Double>, Serializable, PartitionAware<String> {
 
         @Override
         public Double call()
@@ -291,8 +330,9 @@ public class ScheduledExecutorServiceTestSupport extends HazelcastTestSupport {
         }
     }
 
-    public static class HazelcastInstanceAwareRunnable implements Callable<Boolean>, HazelcastInstanceAware, Serializable,
-                                                                  NamedTask {
+    public static class HazelcastInstanceAwareRunnable
+            implements Callable<Boolean>, HazelcastInstanceAware, Serializable, NamedTask {
+
         private transient volatile HazelcastInstance instance;
         private final String name;
 

